design/ceph/ceph-exporter.md

```diff
@@ -0,0 +1,121 @@
+# Ceph exporter
+
+### Goals
+Use Ceph exporter as the new source of metrics based on performance counters coming from every Ceph daemon.
+
+## Use case
+Ceph exporter was created to avoid performance problems at scale obtaining metrics in clusters with high number of daemons (mainly OSDs).
+The Prometheus manager module was not enough performant to obtain the performance counters from all the daemons along the whole Ceph cluster, so the solution for avoid this problem in big clusters is to have a new daemon collecting all the performance counters from all the daemons in each host, and expose them as standard prometheus metrics at /metrics endpoint for prometheus to scrape each of those endpoints.
+### Why do we need this daemon in Rook
+Although probably the Rook clusters usually do not have a huge number of daemons, and therefore, probably the main problem solved by the Ceph Exporter(scalability) is not going to be present, with the Ceph exporter, the source of metrics changes in the Ceph clusters.
+After introducing Ceph Exporter in Ceph cluster we are going to have two main sources of metrics:
+- Prometheus manager module: It is responsible for exposing all metrics other than ceph daemons performance counters.
+- Ceph exporter: It is responsible for exposing only ceph daemons performance counters as prometheus metrics.
+
+So if we do not deploy the Ceph exporter we are not going to have access to the daemon performance counters (in form of metrics) coming from the different Ceph daemons.
+
+## Prerequisites
+- Ceph daemon socket files must have ceph user permission
+- hostPath for ceph daemon socket dirs
+Approach / Implementation details
+
+## Proposal details
+Ceph-exporter is a daemon for collecting perf counters of ceph daemons from admin socket files and exposing them to /metrics endpoint.
+So we will need the following things:
+- Deploy it as a daemonset in all the k8s cluster nodes where a Ceph daemon is running
+- nodeAffinity
+- hostNetwork: true
+- RestartPolicy : Always
+- hostPath for ceph daemon socket dirs
+- Volumes & VolumeMounts for ceph pods & containers for daemon socket files and also ceph config file (/var/lib/rook/rook-ceph/rook-ceph.config)
+- Add daemonSocketDir to ChownCephDataDirsInitContainer()
+- Provide a /metrics endpoint: It needs a new K8s service.
+- Needs to be scraped by prometheus: We need to use a Prometheus Service Monitor which select the Ceph exporter service
+
+#### Example
```

Comment:
It would help to show an example of the services and the service monitor yaml that will be created to help illustrate the behavior (and they are much shorter than a pod spec). For example:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: ceph-exporter-<node>
  namespace: <namespace>
  labels:
    app: rook-ceph-exporter
    node: <node>
spec:
  internalTrafficPolicy: Cluster
  ports:
  - name: http-metrics
    port: 9283
    protocol: TCP
    targetPort: 9283
  selector:
    app: rook-ceph-exporter
    node: <nodeName>
  sessionAffinity: None
  type: ClusterIP
```

And the ServiceMonitor would have the `rook-ceph-exporter` selector (without the `node` selector), so hopefully the single service monitor would collect metrics from all the nodes' services.
```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: rook-ceph-exporter
  namespace: <namespace>
spec:
  namespaceSelector:
    matchNames:
      - <namespace>
  selector:
    matchLabels:
      app: rook-ceph-exporter
  endpoints:
  - port: http-metrics
    path: /metrics
    interval: 5s
```